### DISCUSSION TITLE
Acquiring access token crashes test.

### DISCUSSION THREAD

**Discussion by ErlendLygre** [mswjs/msw#549] [Q&A] [answered]
Hi! 

First of all: I love this library. Great work!

I'm currently implementing MSW in our app. 

We're using Axios for the API-calls, we're using an axios interceptor to acquire an access token before each call. 

Specifically, we're using MSAL 2.0, and running acquireTokenSilently before each request. 

The token request, however, causes the test to crash before it is intercepted by the MSW server. 

Does anyone know a good way to mock/bypass this token acquisition step in tests? 

**MetaTableCell.tsx**
<img width="1027" alt="Screenshot 2021-01-20 at 11 15 21" src="https://user-images.githubusercontent.com/8993868/105160922-159b4100-5b11-11eb-9cd8-f8f25ca39b0f.png">
<img width="639" alt="Screenshot 2021-01-20 at 11 12 51" src="https://user-images.githubusercontent.com/8993868/105160943-1b912200-5b11-11eb-9c10-0f5bd22c17ed.png">

**API.ts. 
On submit, this API endpoint is called.**
<img width="776" alt="Screenshot 2021-01-20 at 11 33 08" src="https://user-images.githubusercontent.com/8993868/105162926-86dbf380-5b13-11eb-8947-4bb96b78ad81.png">

**API.ts
The request is then intercepted by axios, attempting to get an access token**
<img width="688" alt="Screenshot 2021-01-20 at 11 11 00" src="https://user-images.githubusercontent.com/8993868/105161085-43808580-5b11-11eb-8c50-97b5a2f9c870.png">

**authService.ts
Token is requested by MSALs acquireTokenSilently. It needs an account passed into it. This is added to a global object after user logs in. If one is not provided, it will fail and redirect user to login screen.** 
<img width="753" alt="Screenshot 2021-01-20 at 11 20 08" src="https://user-images.githubusercontent.com/8993868/105161320-893d4e00-5b11-11eb-980d-5c60d2de177e.png">

**MetaTableCell.test.tsx**
<img width="1045" alt="Screenshot 2021-01-20 at 11 11 53" src="https://user-images.githubusercontent.com/8993868/105160985-28ae1100-5b11-11eb-9145-d5ca475de469.png">

**setupServer.ts**
<img width="1053" alt="Screenshot 2021-01-20 at 11 10 33" src="https://user-images.githubusercontent.com/8993868/105163384-24372780-5b14-11eb-8990-49b1d4269d52.png">

**The error message. I'm unable to pass a user account as the user is not logged in**
![Screenshot 2021-01-20 at 11 09 59](https://user-images.githubusercontent.com/8993868/105162059-69f2f080-5b12-11eb-9af1-cc4eb0f4c36a.jpg)

**Comment by kettanaito** [comment, marked as answer]
Hey. @ErlendLygre! Thank you for the kind words!

Please, could you share some code snippets or, ideally, a minimal reproduction repo with this issue? That'd help us to get to the bottom of it much more efficiently. 

> The token request, however, causes the test to crash before it is intercepted by the MSW server.

Does the crash happen because of a failed token request, or there's something else? An error stack trace would be good to see.

**Reply by ErlendLygre** [reply]
Yes. I can't really paste it properly with the "insert code"-tool here, but I've uploaded screenshots to the original post now :)

**Reply by ErlendLygre** [reply]
Indeed it happens because of a failed token request. In the normal flow, the user is redirected to login whenever a token request fails.

**Reply by kettanaito** [reply]
Hey, @ErlendLygre. Do you mind marking this discussion as answered? It looks like the issue has been solved. Thanks.

**Comment by ErlendLygre** [comment]
I was able to solve this by mocking the auth service with 

jest.mock('../../../../../services/authService'); 

and creating a __mock__ directory with an authService.ts-file containing 

export const getAccessToken = () => {
    return Promise.resolve('Bearer test');
};

I'm new to mocking, and didn't realize I could mock an import of an import.

**Comment by kettanaito** [comment]
Thanks for sharing the screenshots, they are of great help.

> I'm new to mocking, and didn't realize I could mock an import of an import.

Yes, you can. However, I'd rather you didn't. When you mock an entire `authService` your test no longer calls its code. But your production app will still do that. If you want to get more confidence from your test, you need to let that `authService` to execute, because it's a part of your app's logic. What you should be mocking is any HTTP requests your app or that service makes—those don't belong in a test and make it flaky. 

I'll try to look into your use case once I have a spare minute. Meanwhile, anybody else is welcome to investigate!

**Comment by kettanaito** [comment]
@ErlendLygre, what happens if you remove the `await auth.getAccesToken()` call?

```diff
axios.interceptors.reques.use(async (request) => {
  try {
-   const token = await auth.getAccessToken()
+   const token = 'abc-123'
  }
})
```

Let's simplify the usage example to see if `axios.interceptors` is the issue. The exception above is originating during the auth token retrieval and doesn't look connected to MSW. Let's replace the `token` with a hard-coded value and see what happens.
